packages/contracts-bedrock/src/L2/CrossL2Inbox.sol

```diff
@@ -0,0 +1,139 @@
+
+/// @title IL1Block_OnlyIsInDependencySet
+/// @notice Interface for L1Block with only `isInDependencySet(uint256)` method.
+interface IL1Block_OnlyIsInDependencySet {
+    /// @notice Returns true iff the chain associated with input chain ID is in the interop dependency set.
+    ///         Every chain is in the interop dependency set of itself.
+    /// @param _chainId The input chain ID.
+    /// @return True if the input chain ID corresponds to a chain in the interop dependency set. False otherwise.
+    function isInDependencySet(uint256 _chainId) external view returns (bool);
+}
+
+/// @custom:proxied
+/// @custom:predeploy 0x4200000000000000000000000000000000000022
+/// @title CrossL2Inbox
+/// @notice The CrossL2Inbox is responsible for executing a cross chain message on the destination
+///         chain. It is permissionless to execute a cross chain message on behalf of any user.
+contract CrossL2Inbox is ISemver {
+    /// @notice The struct for a pointer to a message payload in a remote (or local) chain.
+    struct Identifier {
+        address origin;
+        uint256 blocknumber;
+        uint256 logIndex;
+        uint256 timestamp;
+        uint256 chainId;
+    }
+
+    /// @notice Transient storage slot that the origin for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.origin")) - 1)
+    bytes32 public constant ORIGIN_SLOT = 0xd2b7c5071ec59eb3ff0017d703a8ea513a7d0da4779b0dbefe845808c300c815;
+
+    /// @notice Transient storage slot that the blocknumber for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.blocknumber")) - 1)
+    bytes32 public constant BLOCKNUMBER_SLOT = 0x5a1da0738b7fdc60047c07bb519beb02aa32a8619de57e6258da1f1c2e020ccc;
+
+    /// @notice Transient storage slot that the logIndex for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.logindex")) - 1)
+    bytes32 public constant LOG_INDEX_SLOT = 0xab8acc221aecea88a685fabca5b88bf3823b05f335b7b9f721ca7fe3ffb2c30d;
+
+    /// @notice Transient storage slot that the timestamp for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.timestamp")) - 1)
+    bytes32 public constant TIMESTAMP_SLOT = 0x2e148a404a50bb94820b576997fd6450117132387be615e460fa8c5e11777e02;
+
+    /// @notice Transient storage slot that the chainId for an Identifier is stored at.
+    ///         Equal to bytes32(uint256(keccak256("crossl2inbox.identifier.chainid")) - 1)
+    bytes32 public constant CHAINID_SLOT = 0x6e0446e8b5098b8c8193f964f1b567ec3a2bdaeba33d36acb85c1f1d3f92d313;
+
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice Returns the origin address of the Identifier.
+    /// @return _origin The origin address of the Identifier.
+    function origin() external view returns (address _origin) {
```

Comment:
I am wondering if we should revert these calls when called from a callframe not below `executeMessage` to differentiate from the zero values

Reply:
im not sure I understand. could you explain more?

Reply:
Ok so the problem is that `tload(0x20)` will read `bytes(0)` in the case of both `tstore(0x20, 0)` and also nothing being stored there. Its kind of like the difference of a pointer being `nil` vs `0`.

It creates a safety issue where it is not possible to differentiate between 0 and `nil`

The way that we do this is define a magic slot using the digest scheme over "crossl2inbox.entered" and then set a value in it, it doesn't really matter what the value is because its 200 gas no matter what. Then in each of the getters, `tload` the magic slot and if its `0`, revert with the same message. This means `nil` is revert and `0` is legit because no revert